File: src/ClaudeToolbar.Core/Mascot/ThresholdCueTracker.cs
using ClaudeToolbar.Core.Formatting;
using ClaudeToolbar.Core.Usage;
using ClaudeToolbar.Core.Widget;

namespace ClaudeToolbar.Core.Mascot;

public readonly record struct TrackedRow(string Key, BarLevel Level, DateTimeOffset? ResetsAt);

public static class TrackedRows
{
    /// <summary>Pairs each visible widget row with the reset date of the window it shows, keyed by the row label.</summary>
    public static IReadOnlyList<TrackedRow> From(WidgetModel widget, UsageSnapshot? snapshot)
    {
        var rows = new List<TrackedRow>(widget.Rows.Count);
        foreach (var row in widget.Rows)
        {
            var window = row.Label switch
            {
                "5h" => snapshot?.FiveHour,
                "7d" => snapshot?.SevenDay,
                "7d Opus" => snapshot?.SevenDayOpus,
                "7d Sonnet" => snapshot?.SevenDaySonnet,
                _ => null,
            };
            rows.Add(new TrackedRow(row.Label, row.Level, window?.ResetsAt));
        }
        return rows;
    }
}

/// <summary>Emits a cue when a row's level rises within the same reset period. The first observation of a key only primes it.</summary>
public sealed class ThresholdCueTracker
{
    private readonly Dictionary<string, TrackedRow> _last = new();

    public MascotCue? Observe(IEnumerable<TrackedRow> rows)
    {
        MascotCue? cue = null;
        var seen = new HashSet<string>();
        foreach (var row in rows)
        {
            seen.Add(row.Key);
            if (_last.TryGetValue(row.Key, out var previous) && previous.ResetsAt == row.ResetsAt && row.Level > previous.Level)
            {
                var candidate = row.Level == BarLevel.Crit ? MascotCue.Crit : MascotCue.Warn;
                if (cue is null || candidate > cue) cue = candidate;
            }
            _last[row.Key] = row;
        }

        foreach (var stale in _last.Keys.Where(k => !seen.Contains(k)).ToList())
            _last.Remove(stale);

        return cue;
    }

    public void Reset() => _last.Clear();
}

File: mac/Sources/ClaudeToolbarCore/Mascot/ThresholdCueTracker.swift
import Foundation

public struct TrackedRow: Equatable, Sendable {
    public let key: String
    public let level: BarLevel
    public let resetsAt: Date?

    public init(key: String, level: BarLevel, resetsAt: Date?) {
        self.key = key
        self.level = level
        self.resetsAt = resetsAt
    }
}

extension BarLevel {
    /// ok < warn < crit.
    var rank: Int {
        switch self {
        case .ok: return 0
        case .warn: return 1
        case .crit: return 2
        }
    }
}

public enum TrackedRows {
    /// Pairs each visible row with the reset date of the window it shows, keyed by the row label.
    public static func from(_ model: StatusItemModel, snapshot: UsageSnapshot?) -> [TrackedRow] {
        model.rows.map { row in
            let window: UsageWindow?
            switch row.label {
            case "5h": window = snapshot?.fiveHour
            case "7d": window = snapshot?.sevenDay
            case "7d Opus": window = snapshot?.sevenDayOpus
            case "7d Sonnet": window = snapshot?.sevenDaySonnet
            default: window = nil
            }
            return TrackedRow(key: row.label, level: row.level, resetsAt: window?.resetsAt)
        }
    }
}

/// Emits a cue when a row's level rises within the same reset period. The first observation of a key only primes it.
public final class ThresholdCueTracker {
    private var last: [String: TrackedRow] = [:]

    public init() {}

    public func observe(_ rows: [TrackedRow]) -> MascotCue? {
        var cue: MascotCue?
        var seen = Set<String>()
        for row in rows {
            seen.insert(row.key)
            if let previous = last[row.key], previous.resetsAt == row.resetsAt, row.level.rank > previous.level.rank {
                let candidate: MascotCue = row.level == .crit ? .crit : .warn
                if let current = cue {
                    if candidate > current { cue = candidate }
                } else {
                    cue = candidate
                }
            }
            last[row.key] = row
        }
        last = last.filter { seen.contains($0.key) }
        return cue
    }

    public func reset() {
        last.removeAll()
    }
}
